feat(storage): Add SigV4 authentication support for Elasticsearch/OpenSearch storage backends

SoumyaRaikwar opened the pull request:
This PR enables Jaeger to use AWS Managed Elasticsearch/OpenSearch for trace and metrics storage by adding SigV4 HTTP authentication support to Elasticsearch and OpenSearch backends.

## Summary of changes

**Configuration**
- Add `jaeger_storage.backends.<name>.<elasticsearch|opensearch>.auth_extension.authenticator` to reference an OpenTelemetry HTTP authenticator extension by name
- Add `jaeger_storage.metric_backends.<name>.<elasticsearch|opensearch>.auth_extension.authenticator` for metric storage backends

**Elasticsearch/OpenSearch backends**
- Thread the resolved HTTP authenticator through the factory chain (v1/v2 trace storage and metrics storage)
- Wrap the HTTP RoundTripper used by ES/OS clients with the extension's RoundTripper (applies SigV4 signing when using `sigv4authextension` )
- Updated `GetHTTPRoundTripper()` to accept and apply the HTTP authenticator

## Configuration example

```yaml
extensions:
sigv4auth:
region: us-east-1
service: es # or 'aoss' for OpenSearch Serverless
# credentials/assume-role configuration per the extension's documentation

service:
extensions: [sigv4auth]

jaeger_storage:
backends:
es-aws:
elasticsearch:
servers: ["https://my-domain.us-east-1.es.amazonaws.com/"]
auth_extension:
authenticator: sigv4auth
indices:
spans:
shards: 5
replicas: 1

metric_backends:
es-metrics:
elasticsearch:
servers: ["https://my-domain.us-east-1.es.amazonaws.com/"]
auth_extension:
authenticator: sigv4auth
```

## Implementation

- ES/OS backends now support optional HTTP authenticators via `auth_extension.authenticator`
- The extension's RoundTripper wraps the base transport for SigV4 signing
- Supports trace and metrics storage for Elasticsearch 7.x/8.x and OpenSearch

## Scope

- Adds authentication support to:
  - Elasticsearch trace storage (v1 and v2)
  - OpenSearch trace storage (v1 and v2)
  - Elasticsearch metrics storage
  - OpenSearch metrics storage
- Backward compatible - authentication is optional

## Related issue

Part of jaegertracing/jaeger#7468

yurishkuro commented on internal/storage/metricstore/elasticsearch/factory.go:
```diff
@@ -20,12 +22,13 @@ type Factory struct {
 // NewFactory creates a new Factory with the given configuration and telemetry settings.
-func NewFactory(ctx context.Context, cfg config.Configuration, telset telemetry.Settings) (*Factory, error) {
+// Pass nil for httpAuth if authentication is not required.
+func NewFactory(ctx context.Context, cfg config.Configuration, telset telemetry.Settings, httpAuth extensionauth.HTTPClient) (*Factory, error) { // ADD httpAuth parameter
```
```suggestion
func NewFactory(
	ctx context.Context,
	cfg config.Configuration,
	telset telemetry.Settings,
	httpAuth extensionauth.HTTPClient,
) (*Factory, error) {
```

yurishkuro commented on cmd/jaeger/internal/extension/jaegerstorage/extension.go:
can this use s.resolveAuthenticator too?

yurishkuro replied:
is the code pattern here different from other storages? I was expecting to see identical code

SoumyaRaikwar replied:
@yurishkuro Good catch on the pattern consistency. Since Prometheus already has `Auth.Authenticator` (similar to our `AuthExtension.Authenticator`), I can make `resolveAuthenticator` accept both config types using type switching. This would make all three backends (ES/OS/Prometheus) use the identical call pattern:
```
httpAuth, authErr := s.resolveAuthenticator(host, cfg.Backend.Auth, "backend", name)
```
The helper would handle the type differences internally. Does this approach work for you, or would you prefer a different solution?

yurishkuro commented on internal/storage/v1/elasticsearch/factory.go:
```diff
@@ -43,7 +44,7 @@ type FactoryBase struct {
 	tracer         trace.TracerProvider
 
-	newClientFn func(ctx context.Context, c *config.Configuration, logger *zap.Logger, metricsFactory metrics.Factory) (es.Client, error)
+	newClientFn func(ctx context.Context, c *config.Configuration, logger *zap.Logger, metricsFactory metrics.Factory, httpAuth extensionauth.HTTPClient) (es.Client, error) // ADD httpAuth parameter
```
```suggestion
	newClientFn func(ctx context.Context, c *config.Configuration, logger *zap.Logger, metricsFactory metrics.Factory, httpAuth extensionauth.HTTPClient) (es.Client, error)
```

yurishkuro commented on internal/storage/v1/elasticsearch/helper.go:
```diff
@@ -21,7 +22,8 @@ type mockClientBuilder struct {
 }
 
-func (m *mockClientBuilder) NewClient(context.Context, *escfg.Configuration, *zap.Logger, metrics.Factory) (es.Client, error) {
+// UPDATE THIS FUNCTION - Add httpAuth parameter
```
```suggestion
```

SoumyaRaikwar commented on internal/storage/elasticsearch/config/config.go:
```diff
@@ -99,6 +100,9 @@ type Configuration struct {
 	Authentication     Authentication `mapstructure:"auth"`
+	// AuthExtension configures HTTP authentication using OpenTelemetry extensions
+	// (e.g., sigv4auth for AWS). This is independent of the Authentication field above.
+	AuthExtension *AuthExtensionConfig `mapstructure:"auth_extension,omitempty"`
```
done! Moved `AuthExtension` into the `Authentication` struct for better organization.

yurishkuro commented on cmd/jaeger/internal/extension/jaegerstorage/config.go:
```diff
@@ -68,10 +68,14 @@ type AuthConfig struct {
 type PrometheusConfiguration struct {
 	promcfg.Configuration `mapstructure:",squash"`
-	Auth                  *AuthConfig `mapstructure:"auth,omitempty"`
+	Authentication        AuthenticationConfig `mapstructure:"authentication"`
```
this is not backwards compatible change

SoumyaRaikwar replied:
reverted it

yurishkuro commented on internal/storage/elasticsearch/config/config.go:
```diff
@@ -208,6 +209,21 @@ type Authentication struct {
 	BasicAuthentication configoptional.Optional[BasicAuthentication] `mapstructure:"basic"`
 	BearerTokenAuth     configoptional.Optional[TokenAuthentication] `mapstructure:"bearer_token"`
 	APIKeyAuth          configoptional.Optional[TokenAuthentication] `mapstructure:"api_key"`
+	AuthExtension       *AuthExtensionConfig                         `mapstructure:"auth_extension,omitempty"`
```
what is the point of this new class? Why not use `configauth.Config` from OTEL?

SoumyaRaikwar replied:
I've switched to using `configauth.Config` from OTEL as a type alias. This eliminates the custom struct while maintaining the same functionality.

yurishkuro commented on internal/storage/elasticsearch/config/config.go:
```diff
@@ -210,7 +211,7 @@ type Authentication struct {
 	BearerTokenAuth     configoptional.Optional[TokenAuthentication] `mapstructure:"bearer_token"`
 	APIKeyAuth          configoptional.Optional[TokenAuthentication] `mapstructure:"api_key"`
-	AuthExtension       *AuthExtensionConfig                         `mapstructure:"auth_extension,omitempty"`
+	Authenticator       component.ID                                 `mapstructure:"authenticator"`
```
```suggestion
	configauth.Config `mapstructure:",squash"`
```

SoumyaRaikwar commented:
Hi @yurishkuro,

I've implemented your suggestion to embed `configauth.Config` in the `Authentication` struct:
```
type Authentication struct {
BasicAuthentication configoptional.Optional[BasicAuthentication] mapstructure:"basic"
BearerTokenAuth configoptional.Optional[TokenAuthentication] mapstructure:"bearer_token"
APIKeyAuth configoptional.Optional[TokenAuthentication] mapstructure:"api_key"
configauth.Config mapstructure:",squash"
}
```

However, I'm encountering a compiler error: `unknown field AuthenticatorID in struct literal` in `config_test.go`, even after clearing Go caches and rebuilding.

Could you advise if I'm missing something in how the embedded field should be exposed, or if there's a different approach I should take?

Thanks!

yurishkuro commented on cmd/jaeger/internal/extension/jaegerstorage/extension.go:
```diff
@@ -184,19 +185,33 @@ func (s *storageExt) Start(ctx context.Context, host component.Host) error {
+					AuthenticatorID: cfg.Elasticsearch.Authentication.AuthenticatorID,
+				}
+			}
+			httpAuth, authErr := s.resolveAuthenticator(host, esAuthCfg, "elasticsearch", storageName)
```
too much boilerplate that can be handled by the helper function

```suggestion
			httpAuth, authErr := s.resolveAuthenticator(host, cfg.Elasticsearch.Authentication, "elasticsearch", storageName)
```

yurishkuro commented on internal/storage/elasticsearch/config/config.go:
```diff
@@ -208,8 +210,11 @@ type Authentication struct {
 }
 
+type AuthExtensionConfig = configauth.Config
+
```
```suggestion
```

SoumyaRaikwar commented:
@yurishkuro Changes implemented! 

Removed the `AuthExtensionConfig`  and simplified the code as requested:

- Pass `Authentication` struct directly to `resolveAuthenticator`
- Eliminated all boilerplate wrapping code

The embedded `configauth.Config` with `mapstructure:",squash"` ensures the config structure matches OTEL pattern.

yurishkuro commented on internal/storage/elasticsearch/config/config_test.go:
```diff
@@ -1470,6 +1472,61 @@ func TestBulkCallbackInvoke_NilResponse(t *testing.T) {
 	)
 }
 
+func TestAuthExtensionConfig(t *testing.T) {
```
I don't see the need for this test. First, wantErr is always false, but more importantly what is it actually validating? This package has no knowledge of "extensions".

yurishkuro reviewed (commented):
the code coverage gaps in cmd/jaeger/internal/extension/jaegerstorage/extension.go look relatively easy to fix

SoumyaRaikwar commented:
@yurishkuro Added 7 tests for authenticator error paths (5 in extension_test.go, 2 in config_test.go) to address codecov gaps.
